### PR TITLE
Namespaces in addon update

### PR DESCRIPTION
**Pull request by marekbiskup**
Two PRs had to be cherry-picked to make the test work:
1. ~~~#10609~~~ (#10453 superseded this)
2. #10237

Not tested with @satnam6502 's (and @brendanburns 's) patch yet (I'm going to do it now).
Can be merged before #10487.

It is urgent because it blocks #10487 
I'll be available in the evening to make some fixes, if needed.

### CODE REVIEW

**Comment by marekbiskup** [comment]
cc @zmerlynn

**Comment by zmerlynn** [comment]
I'm going to review the last commit alone and hope that we can get the cherry-picks pushed separately this morning. I assumed you put this up together so you could show it as a testable-unit for e.g. the pull request builder, not because you wanted to push them all at once necessarily.

**Comment by marekbiskup** [comment]
@zmerlynn, that's right.

**Comment by marekbiskup** [comment]
rebased. And I also corrected the problem with IFS in bash 'read'

**Comment by marekbiskup** [comment]
I don't have time to test addon-update e2e after rebase. I'll try to do it in the evening.

**Comment by satnam6502** [comment]
I intend to cherry-pick the commits in the PR (or base them) against my namespace change PR and then run e2e so it should also get tested as part of that.

**Comment by zmerlynn** [comment]
@marekbiskup: I should've done it earlier, but if you have time, I just merged the other of your cherry-picks.

**Comment by zmerlynn** [comment]
LGTM

**Comment by zmerlynn** [comment]
I've confirmed manually that the only thing blocking the rebase is the cherry-pick we've already done. (I ran my "test_pull" script, then rebased to master and it rebased clean.)

cc @quinton-hoole for an ok-to-merge

I'd like to get this ready to merge for when @marekbiskup comes back online, since we're borrowing his evening time.

**Comment by quinton-hoole** [comment]
Wow. Large one.  I'm taking a look now.

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh [hunk 1]
```diff
@@ -45,16 +45,19 @@
...
     exit 1
 fi
 
+# If an add-on definition is incorrect, or a definition has just disappeared
+# from the local directory, the script will still keep on retrying.
+# The script does not end until all retries are done, so
+# one invalid manifest may block updates of other add-ons.
+# Be careful how you set these parameters
+NUM_TRIES=1    # will be updated based on input parameters
+DELAY_AFTER_ERROR_SEC=${TEST_DELAY_AFTER_ERROR_SEC:=10}
```
Comment (quinton-hoole) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Why the TEST_ here?  This is not a test?  Also TEST_DELAY_AFTER_ERROR_SEC is a global environment variable, right? Can we name it something that makes it clear what it means in a global context, e.g. KUBE_ADDON_UPDATE_RETRY_SEC?

Comment (marekbiskup) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: @quinton-hoole , this is a pattern for passing values from tests to scripts. TEST_ prefix means that it is not intended for users of the script to change it. In tests we often want to reduce various timeouts and delays for faster execution.
I'm not actually sure it is used in any script, but anyway, I like this pattern.

**Comment by zmerlynn** [comment]
@quinton-hoole: Did you have any other comments? If not I'd like to merge this ASAP

File: cluster/saltbase/salt/kube-addons/kube-addon-update.sh (hunk 1 above)
Comment (quinton-hoole) [inline cluster/saltbase/salt/kube-addons/kube-addon-update.sh]: Aah - that makes sense.

**Comment by zmerlynn** [comment]
(Obviously after rebase sorted.)

**Comment by marekbiskup** [comment]
I'm just trying to rebase. I'm stuck at checking out the remote branch on my laptop. I aways end up in "detached head state".
Any quick hint?

**Comment by marekbiskup** [comment]
Ok, I missed git pull. Rebasing.

**Comment by quinton-hoole** [comment]
Have you done a local commit?  You'll need to do that.

**Comment by quinton-hoole** [comment]
This code looks fine to me, but it's quite a lot to add so soon before v1.0.  
I'm going to suggest that you merge it, but please keep a close eye on the e2e tests and be sure to roll back if it causes any problems.

**Comment by quinton-hoole** [comment]
LGTM, ok-to-merge

**Comment by marekbiskup** [comment]
please wait with merge until I run addon-update test.

**Comment by zmerlynn** [comment]
I'll be watching it. This was stage 2/3 of higher risk stuff.

**Comment by zmerlynn** [comment]
@marekbiskup: Acknowledged.

**Comment by marekbiskup** [comment]
Sorry, I'm unable to build this branch on my mac - something's wrong with cadvisor.
Can someone else try to run Addon-update test, please?

**Comment by zmerlynn** [comment]
@marekbiskup: on it.

**Comment by marekbiskup** [comment]
ok, it helped when I run 'make clean'. I'll try to continue.

**Comment by satnam6502** [comment]
Also I have this folded in my PR so I will first run an e2e just for the addon test and report back and then run the full test suite.

**Comment by marekbiskup** [comment]
Ok, confirmed, test passed. Please merge.

Good luck @satnam6502 with  #10487 !

**Comment by zmerlynn** [comment]
Thanks for confirmation. My cluster just came up with your PR, I was about to test it myself.

**Pull request merged**